lib/srv/db/cloud/watchers/rdsproxy.go

```diff
@@ -0,0 +1,178 @@
+	"github.com/aws/aws-sdk-go/service/rds/rdsiface"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/services"
+	"github.com/gravitational/teleport/lib/srv/db/common"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// rdsDBProxyFetcher retrieves RDS proxies and proxy endpoints.
+type rdsDBProxyFetcher struct {
+	cfg rdsFetcherConfig
+	log logrus.FieldLogger
+}
+
+// newRDSDBProxyFetcher returns a new RDS proxy fetcher instance.
+func newRDSDBProxyFetcher(config rdsFetcherConfig) (Fetcher, error) {
+	if err := config.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &rdsDBProxyFetcher{
+		cfg: config,
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: "watch:rdsproxy",
+			"labels":        config.Labels,
+			"region":        config.Region,
+		}),
+	}, nil
+}
+
+// Get returns RDS proxies and prosy endpoints matching the watcher's
+// selectors.
+func (f *rdsDBProxyFetcher) Get(ctx context.Context) (types.Databases, error) {
+	databases, err := f.getRDSProxyDatabases(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var result types.Databases
+	for _, database := range databases {
+		match, _, err := services.MatchLabels(f.cfg.Labels, database.GetAllLabels())
+		if err != nil {
+			f.log.Warnf("Failed to match %v against selector: %v.", database, err)
+		} else if match {
+			result = append(result, database)
+		} else {
+			f.log.Debugf("%v doesn't match selector.", database)
+		}
+	}
+	return result, nil
+}
+
+// getRDSProxyDatabases returns a list of database resources representing RDS
+// proxies and proxy endpoints.
+func (f *rdsDBProxyFetcher) getRDSProxyDatabases(ctx context.Context) (types.Databases, error) {
+	rdsProxies, err := getRDSProxies(ctx, f.cfg.RDS, maxPages)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	rdsProxyEndpoints, err := getRDSProxyEndpoints(ctx, f.cfg.RDS, maxPages)
+	if err != nil {
+		f.log.Debugf("Failed to get RDS proxy endpoints: %v.", err)
+	}
+
+	databases := types.Databases{}
+	for _, dbProxy := range rdsProxies {
+		// rds.DBProxy has no port information. An extra SDK call is made to
+		// find the port from its targets.
+		port, err := getRDSProxyTargetPort(ctx, f.cfg.RDS, dbProxy.DBProxyName)
+		if err != nil {
+			f.log.Debugf("Failed to get port for RDS proxy %v: %v.", aws.StringValue(dbProxy.DBProxyName), err)
+			continue
+		}
+
+		// Add a database from RDS proxy (default endpoint).
+		database, err := services.NewDatabaseFromRDSProxy(dbProxy, port)
+		if err != nil {
+			f.log.Debugf("Could not convert RDS proxy %q to database resource: %v.",
+				aws.StringValue(dbProxy.DBProxyName), err)
+		} else {
+			databases = append(databases, database)
+		}
+
+		// Add additional proxy endpoints from the same proxy.
+		for key, dbProxyEndpoint := range rdsProxyEndpoints {
+			if aws.StringValue(dbProxyEndpoint.DBProxyName) != aws.StringValue(dbProxy.DBProxyName) {
```

Comment:
I'm not sure I understand why this check is needed TBH. I.e. why do we need to match an RDS endpoint to a particular RDS proxy. Can there be an RDS endpoint that is not attached to any RDS proxy?

Reply:
Every RDS proxy endpoint belongs to one of the RDS proxies. And each RDS proxy endpoint needs two things from their parent RDS proxy, for connection:

- the parent RDS proxy's resource ID is needed in the IAM proxy for rds connect
- the parent RDS proxy's default endpoint is used as the TLS server name because of cert issue.

Sorry for the confusion. I will add more comments somewhere to clarify.